launchable/commands/record/build.py

```diff
@@ -169,100 +212,127 @@ def build(ctx: click.core.Context, build_name: str, source: List[str], max_days:
                     commit_hash = matched.group('hash')
                     name = matched.group('name')
                     if commit_hash and name:
-                        submodules.append((repo_name + "/" + name, repo_dist + "/" + name, commit_hash))
+                        r.append(Workspace(
+                            name=w.name + "/" + name,
+                            dir=w.dir + "/" + name,
+                            commit_hash=commit_hash))
+        return r
+
+    # figure out the commit hash and branch of those workspaces
+    def compute_hash_and_branch(ws: List[Workspace]):
+        # figure out w.commit_hash and w.branch
+        for w in ws:
+            try:
+                if not w.commit_hash:
+                    w.commit_hash = subprocess.check_output("git rev-parse HEAD".split(), cwd=w.dir).decode().replace("\n", "")
+            except Exception as e:
+                click.echo(
+                    click.style(
+                        "Can't get commit hash for {}. Do you run command under git-controlled directory? "
+                        "If not, please set a directory use by --source option.".format(w.dir),
+                        fg='yellow'),
+                    err=True)
+                print(e, file=sys.stderr)
+                sys.exit(1)
+            w.calc_branch_name()
+
+        # Our historical behaviour is to ignore --branch options. In that case, we should reject if that option is present
+        # I also think it's OK to allow this option to explicitly let the branch name set, since the inference
+        #  of the branch name from the workspace is inherently a little fragile
 
-    if len(commits) != 0:
-        invalid = False
-        _commits = normalize_key_value_types(commits)
+    # Rely on --commit and --branch to create a list of workspaces, even when there's no local Git workspaces
+    def synthesize_workspaces() -> List[Workspace]:
+        # workspace by its name
+        ws = {}  # type: Dict[str,Workspace]
 
         commit_pattern = re.compile("[0-9A-Fa-f]{5,40}$")
-        for repo_name, hash in _commits:
+
+        branch_name_map = dict(normalize_key_value_types(branches))
```

Comment:
I'm scratching my head with `normalize_key_value_types`. Shouldn't our key/value style option handling be consistent across the board? Our intended mechanism to do so is `KeyValueType` in `utils/click.py`, so I'd love to have better comment in the code to explain the motivation behind `normalize_key_value_types`, and why that should only apply to options in this command, but not similar options elsewhere.